pkg/controller/attach_detach.go

```diff
@@ -108,8 +108,14 @@ func (r *ReconcileAttachDetach) Reconcile(ctx context.Context, request reconcile
 		return reconcileReturnOnSuccess(azVolumeAttachment.Name, r.retryInfo)
 	}
 
-	// detachment request
+	// deletion request
 	if objectDeletionRequested(azVolumeAttachment) {
+		if err := r.removeFinalizer(ctx, azVolumeAttachment); err != nil {
+			return reconcileReturnOnError(ctx, azVolumeAttachment, "delete", err, r.retryInfo)
+		}
+	}
+	// detachment request
+	if volumeDetachRequested(azVolumeAttachment) {
```

Comment:
This should be an `else if`. Since the object deletion happens *after* the detach completes, we don't need to process a second detach for the `AzVolumeAttachment`.